src/libsystemd/sd-json/json-util.c

```diff
@@ -281,3 +281,47 @@ int json_dispatch_pidref(const char *name, sd_json_variant *variant, sd_json_dis
 
         return 0;
 }
+
+int json_variant_new_devnum(sd_json_variant **ret, dev_t devnum) {
+        if (devnum == 0)
+                return sd_json_variant_new_null(ret);
+
+        return sd_json_buildo(
+                        ret,
+                        SD_JSON_BUILD_PAIR_UNSIGNED("major", major(devnum)),
+                        SD_JSON_BUILD_PAIR_UNSIGNED("minor", minor(devnum)));
+}
+
+int json_dispatch_devnum(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        dev_t *ret = ASSERT_PTR(userdata);
+        int r;
+
+        assert(variant);
+
+        if (sd_json_variant_is_null(variant)) {
+                *ret = 0;
+                return 0;
+        }
+
+        struct {
+                uint32_t major;
+                uint32_t minor;
+        } data;
+
+        static const sd_json_dispatch_field dispatch_table[] = {
+                { "major", _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint32, voffsetof(data, major), SD_JSON_MANDATORY },
+                { "minor", _SD_JSON_VARIANT_TYPE_INVALID, sd_json_dispatch_uint32, voffsetof(data, minor), SD_JSON_MANDATORY },
+                {}
+        };
+
+        r = sd_json_dispatch(variant, dispatch_table, flags, &data);
+        if (r < 0)
+                return r;
+
+        dev_t devnum = makedev(data.major, data.minor);
+        if (major(devnum) != data.major || minor(devnum) != data.minor)
```

Comment:
hmm, glibc major/minor space is artificially large, larger than what Linux actually supports. Hence these tests would lie about what is actually valid. Please use DEVICE_MAJOR_VALID() and DEVICE_MINOR_VALID() for validation instead